### PR TITLE
RSDK-4894: Correct locking around remote logging creation of grpc clients.

### PR DESCRIPTION
**Pull request by dgottlieb**

### CODE REVIEW

File: web/server/net_logger.go [hunk 1]
```diff
@@ -328,22 +331,27 @@ func (w *remoteLogWriterGRPC) write(logs []*apppb.LogEntry) error {
...
+	w.rpcClient = client
+	w.service = apppb.NewRobotServiceClient(w.rpcClient)
 	return w.service, nil
 }
 
 func (w *remoteLogWriterGRPC) close() {
+	w.clientMutex.Lock()
+	defer w.clientMutex.Unlock()
+
 	if w.rpcClient != nil {
```
Comment (dgottlieb) [inline web/server/net_logger.go:355]: My reading of the code suggests this is a perfectly synchronized read of the `w.rpcClient` member that won't race with a write. But one can't come to that conclusion by only reading the `remoteLogWriter` implementation. So I thought it best to make the implementation obviously correct from a data race perspective and resilient to other code changing which makes this incorrect.

Certainly if close is racing with writers trying to send over the client -- we're in a bad place anyways. But similarly I don't see a need to poke the data race bear.

File: web/server/net_logger.go [hunk 2]
```diff
@@ -298,7 +298,10 @@ type remoteLogWriter interface {
 }
 
 type remoteLogWriterGRPC struct {
-	cfg         *config.Cloud
+	cfg *config.Cloud
+
+	// `service` and `rpcClient` are lazily initialized on first call to `write`. The `clientMutex`
+	// serializes access to ensure that only one caller creates them.
```
Comment (benjirewis) [inline web/server/net_logger.go:304]: Nice 👍🏻

File: web/server/net_logger.go [hunk 3]
```diff
@@ -328,22 +331,27 @@ func (w *remoteLogWriterGRPC) write(logs []*apppb.LogEntry) error {
 }
 
 func (w *remoteLogWriterGRPC) getOrCreateClient(ctx context.Context) (apppb.RobotServiceClient, error) {
-	if w.service == nil {
-		w.clientMutex.Lock()
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply web/server/net_logger.go:332]: Agree!

File: web/server/net_logger.go (hunk 1 above)
Comment (benjirewis) [reply web/server/net_logger.go:355]: I think this lock is wise to put in despite the attempt to clear the log queue in `Close`.

**Review by benjirewis — APPROVED** [review]
Thanks!

**Pull request merged**